libraries/plugins/es_objects/es_objects.cpp

```diff
@@ -141,61 +147,126 @@ bool es_objects_plugin_impl::index_database(const vector<object_id_type>& ids, s
          limit_documents = _es_objects_bulk_replay;
 
 
-      for (auto const &value: ids) {
-         if (value.is<proposal_object>() && _es_objects_proposals) {
-            auto obj = db.find_object(value);
-            auto p = static_cast<const proposal_object *>(obj);
-            if (p != nullptr) {
-               if (action == "delete")
-                  remove_from_database(p->id, "proposal");
-               else
-                  prepareTemplate<proposal_object>(*p, "proposal");
+      for (auto const &value: ids)
+      {
+         switch( value.space_type() )
```

Comment:
> Don't know how to do it better though. `object_id_type` can't derive the object type from space and type, so visitor pattern can't be applied.

Where can I look up the visitor pattern in the core code? And what exactly does it in this context?

Reply:
The visitor-pattern is the object-oriented version of a switch statement that switches depending on the type of an object. It is often used when dealing with `static_variant` objects, specifically operations. See e. g. `graphene::chain::operation_get_impacted_accounts` in `db_notify.cpp`.
Since `object_id_type` doesn't know its object class the pattern cannot be applied.

Reply:
Btw, `get_relevant_accounts` in `db_notify.cpp` also uses if+switch, so I guess it's ok for now.